File: code/fenwick.py
class Fenwick:
    def __init__(self, n: int):
        self.n = n
        self.bit = [0]*(n+1)
    def update(self, id: int, val: int):
        while id <= self.n:
            self.bit[id] += val
            id += id&-id
    def query(self, id: int) -> int:
        sum = 0
        while id > 0:
            sum += self.bit[id]
            id -= id&-id
        return sum

File: code/fenwick.cpp
template <typename T>
struct Fenwick {
	vector<T> bit;

	Fenwick(int sz): sz(sz), bit(sz+1) {}

	void update(int id, T val) {
		for(;id<(int)bit.size();id+=id&-id) bit[id]+=val;
	}

	T query(int id) {
		T sum = 0;
		for(;id>0;id-=id&-id) sum+=bit[id];
		return sum;
	}
};
